### PR TITLE
feature: add support for parallel processor

### LINKED ISSUE #2659: add support for parallel analysis
for pure read-only analysis we can run the processors on multiple cores (one AST node per core)

### PR DESCRIPTION
**Pull request by MartinWitt**
Fix #2659

The new class allows parallel processing for stateless processors. Subtyping processor allows easy usage in launcher and keeps it simple for the user.
Currently the are still some problems and needed changes:

- [x] writing doc ( and testing)
- [X] easier usage => added constructor with Consumer as argument
- [X] allow usage for non stateless processors => some problems where you need a shared state you could easily share the state between processors by collections. Some problems are never parallel doable. so some work for the user will be left.

wdyt about it?

### CODE REVIEW

**Comment by monperrus** [comment]
I think it's really cool!

**Comment by MartinWitt** [comment]
i think the feature is finished if no one sees any problems?

**Comment by monperrus** [comment]
Great progress.

In Spoon, each time we add a feature, we document it in the `doc` folder. Could you document the parallel processor in `processor.md`?

In Spoon, we try to add a line `// contract: the tested contract in natural language` in each test case. Could you do so?

Thanks!

File: src/main/java/spoon/processing/AbstractParallelProcessor.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	/**
+	 * Creates a new AbstractParallelProcessor from given consumer. The processors
+	 * must follow the guidelines given in the class description.
+	 *
+	 * @param processFunction    Represents an operation that accepts a single
+	 *                           element E and returns no result.
+	 * @param numberOfProcessors number of concurrent running processors.
+	 * @throws IllegalArgumentException if numberOfProcessors is less than 1.
+	 */
+	public AbstractParallelProcessor(Consumer<E> processFunction, int numberOfProcessors) {
```
Comment (monperrus) [inline src/main/java/spoon/processing/AbstractParallelProcessor.java]: neat API!

Comment (MartinWitt) [reply src/main/java/spoon/processing/AbstractParallelProcessor.java]: thanks

**Comment by MartinWitt** [comment]
For both failing test cases see #3289 (?) and wait with merges till it is resolved?

**Comment by monperrus** [comment]
Thanks a lot @MartinWitt

**Comment by monperrus** [comment]
For the record, #1111 is related to this.

**Pull request merged**